spec/src/main/asciidoc/microprofile-lra-spec.adoc

```diff
@@ -408,6 +408,12 @@ executing the `@LRA` annotated method, the implementation must be able to guaran
 participant has been or will eventually be registered with the LRA.
 This implies that the LRA is effectively validated as being active (since
 it is not possible to enlist with an inactive LRA).
+footnoteref:[note1,there is a caveat for nested LRAs which have an extra state model transition
+which allows an inactive LRA to transition from the `Closed` state to the `Cancelling` state.
+To facilitate this transition we allow enlistement if the implementation knows that the incoming
+context is nested and in the state `Closed` or `Active`. And in the case where the LRA is `Closed`,
+if the method finishes without error, the complete method, if present, would be invoked as normal.]
+
```

Comment:
Sorry, this footnote to me still causes more confusion then it explains the intended behavior. For instance, is `return Response.status(4xx).build()` finish with or without error?

In the issue you mentioned:

if the implementation knows that the incoming context is nested and not cancelled then we allow enlistment (even if it is closed), otherwise the method is not allowed to proceed.

Can you incorporate rather this information, please?

Reply:
And if the Complete method can be invoked again should we mark this transition in the state diagram? What happens if the nested Complete method on repeated close of nested LRA transforms nested LRA from Closed to FailedToClose state?

Reply:
The text `if the method finishes without error` means that the method does not result in a status code that causes the LRA to cancel: HTTP error statuses in the context of LRA mean to transition to Cancelling (ie the ones in cancelOn, cancelOnFamily attributes). I thought that was implicit but I can change it something more explicit.

To your second comment about `the Complete method can be invoked again ...`: the complete method is not invoked a second time so the transition cannot occur (for nested LRAs, only Closed -> Cancelling is permitted which would result in the Compensate method being called). The text you quote is referring to a new participant joining a closed nested LRA so its complete method would only be called once.

Reply:
> the method does not result in a status code that causes the LRA to cancel

This would be better so please change it to something more explicit. Don't forget also about non-JAX-RS cases.

>  The text you quote is referring to a new participant joining a closed nested LRA so its complete method would only be called once.

Ok, thanks. I understand now. But still, this new participant can return FailedToComplete from its Complete method invocation and thus there is a transition Closed -> FailedToClose or it can return Closing and thus there is the transition Closed -> Closing?

Reply:
This is a great observation. I do not think we should start messing up the state model with complex transitions so let's either:

1. Prohibit new participants from joining closed nested LRAs, or
2. Prohibit new participants with an `@Complete` callback from joining closed nested LRAs,

or both.

In the first situation the nested LRA can be cancelled by invoking a method on the original participant which returns a status code that results in cancel.

In the second situation the nested LRA can be cancelled by invoking a method on the original participant which returns a status code that results in cancel.

Reply:
FYI I just updated my last comment which had a typo in the last paragraph (I replaced `original` with `new`).

Reply:
I would vote for both. Only already enlisted nested participant LRA method can be invoked again and if this method closes LRA the Complete methods are not replayed again (thus no-op). If the LRA method cancels the nested LRA is moved to Cancelling state.

New participants are rejected with 412. This also confirms with the statement that we don't allow enlistment with already closed LRA.

Reply:
Agreed. This approach keeps the model relatively simple (and it means that the state models for LRAs and Participants are kept in harmony). I will update the proposal to include these two options and also include the text you proposed in the last comment.

In addition, your earlier comment about missing transitions prompted me to revisit the participant model. Would you agree that the participant model needs the transition `Completed -> Compensating` to match the LRA state model transition `Closed -> Cancelling` when the LRA is nested. If so then I will also update `Figure 3. Participant State Model` in this PR.

Reply:
> Would you agree that the participant model needs the transition Completed -> Compensating to match the LRA state model transition Closed -> Cancelling when the LRA is nested. If so then I will also update Figure 3. Participant State Model in this PR.

Yes, this model should also be updated.

Reply:
my bad, there is already an issue open to update the model - https://github.com/eclipse/microprofile-lra/issues/277. So this PR is good to go in my opinion. Only waiting for approval from @rdebusscher.

Reply:
> Don't forget also about non-JAX-RS cases.

This text refers to methods annotated with @LRA but I will note your comment when updating the rest of this PR.

Reply:
I see that you decided to rather prohibit the Complete invocation in case the nested LRA method closes the nested LRA again. I don't see a problem with this approach, just documenting it here in case someone reads this thread again in the future.